### PR TITLE
Discover and check for rogue print statements

### LINKED ISSUE #504: detect print statements
**Is your feature request related to a problem? Please describe.**
I want to prevent rogue print statements, like it's currently done for browser()

**Describe the solution you'd like**
Refuse print statements

**Describe alternatives you've considered**
End up with tons of unwanted garbage in the logs :)

**Additional context**
Trivial to implement. I might open a PR

### PR DESCRIPTION
**Pull request by stefanoborini**
Closes #504

### CODE REVIEW

**Review by lorenzwalthert — COMMENTED** [review]
Thanks @stefanoborini. This looks pretty good to me. Thanks for taking the time to make such a complete PR. I am not sure this hook should be enabled by default for projects, since it's common practice to use print statements in analysis / one off projects. An alternative to checking for `print()` with pre-commit hooks would be to implement such a check in {lintr} and then use the lintr pre-commit hook. Did you consider that?

**Comment by stefanoborini** [comment]
> Thanks @stefanoborini. This looks pretty good to me. Thanks for taking the time to make such a complete PR.

Well, I simply copied browser() ;)


> I am not sure this hook should be enabled by default for projects

Absolutely not. But I develop quite complex shiny apps and I've been burned with tons of garbage in the production deployment logs already twice (meaning I have to redeploy etc), so it might be useful for those who, like me, don't want to accidentally forget this kind of rogue statements.

> since it's common practice to use print statements in analysis / one off projects. An alternative to checking for `print()` with pre-commit hooks would be to implement such a check in {lintr} and then use the lintr pre-commit hook. Did you consider that?

No but lintr is on r-lib, and I don't want to collaborate with them.

File: vignettes/available-hooks.Rmd [hunk 1]
```diff
@@ -146,6 +146,13 @@ Guarantees you that you don't accidentally commit code with a
 
 This hook does not modify files.
 
+## `no-print-statement`
+
+Guarantees you that you don't accidentally commit code with a
+`print()` statement in it.
+
+This hook does not modify files.
```
[the remark this answers is not in the document]
Comment (stefanoborini) [reply vignettes/available-hooks.Rmd:154]: Will do, however please note that I don't fully understand how the system works. Be on the lookout for things that I might have missed. I'll do a commit in a few minutes, and try to fix the failures as well.

**Comment by lorenzwalthert** [comment]
> I am not sure this hook should be enabled by default for projects

>> Absolutely not.

Then can you please remove it from the config template for projects in `/inst/pre-commit-config-proj.yaml`?

And fix the end-2-end test? The other test fails expectedly.

File: vignettes/available-hooks.Rmd (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanoborini) [reply vignettes/available-hooks.Rmd]: Committed. I assumed that the current DESCRIPTION version is the one that is "to be released", given that 9020 is not in the tags at the moment.

**Review by lorenzwalthert — APPROVED** [review]
Great, thanks :D

**Comment by lorenzwalthert** [comment]
@stefanoborini you have to use the `pre-commit autoupdate —bleeding` or similar to set the `rev` in `.pre-commit-config.yaml` to the latest HEAD so you can use the hook that you contributed. I will at some point create a new release / git tag  that autoupdate will fetch without the bleeding edge argument.

**Pull request merged**